odl/test/space/npy_tensors_test.py

```diff
@@ -0,0 +1,1545 @@
+
+
+def test_unary_ops(tspace):
+    """Verify that the unary operators (`+x` and `-x`) work as expected."""
+    for op in [operator.pos, operator.neg]:
+        x_arr, x = noise_elements(tspace)
+
+        y_arr = op(x_arr)
+        y = op(x)
+
+        assert all_almost_equal([x, y], [x_arr, y_arr])
+
+
+def test_scalar_operator(tspace, arithmetic_op):
+    """Verify binary operations with scalars.
+
+    Verifies that the statement y = op(x, scalar) gives equivalent results
+    to NumPy.
+    """
+    if arithmetic_op in (operator.truediv, operator.itruediv):
+        places = int(-np.log10(np.finfo(tspace.dtype).resolution) // 2)
+    else:
+        places = int(-np.log10(np.finfo(tspace.dtype).resolution))
+
+    for scalar in [-31.2, -1, 0, 1, 2.13]:
+        x_arr, x = noise_elements(tspace)
+
+        # Left op
+        if scalar == 0 and arithmetic_op in [operator.truediv,
+                                             operator.itruediv]:
+            # Check for correct zero division behaviour
+            with pytest.raises(ZeroDivisionError):
+                y = arithmetic_op(x, scalar)
+        else:
+            y_arr = arithmetic_op(x_arr, scalar)
+            y = arithmetic_op(x, scalar)
+
+            assert all_almost_equal([x, y], [x_arr, y_arr], places=places)
+
+        # right op
+        x_arr, x = noise_elements(tspace)
+
+        y_arr = arithmetic_op(scalar, x_arr)
+        y = arithmetic_op(scalar, x)
+
+        assert all_almost_equal([x, y], [x_arr, y_arr], places=places)
+
+
+def test_binary_operator(tspace, arithmetic_op):
+    """Verify binary operations with tensors.
+
+    Verifies that the statement z = op(x, y) gives equivalent results
+    to NumPy.
+    """
+    if arithmetic_op in (operator.truediv, operator.itruediv):
+        places = int(-np.log10(np.finfo(tspace.dtype).resolution) // 2)
+    else:
+        places = int(-np.log10(np.finfo(tspace.dtype).resolution))
+
+    [x_arr, y_arr], [x, y] = noise_elements(tspace, 2)
+
+    # non-aliased left
+    z_arr = arithmetic_op(x_arr, y_arr)
+    z = arithmetic_op(x, y)
+
+    assert all_almost_equal([x, y, z], [x_arr, y_arr, z_arr], places=places)
+
+    # non-aliased right
+    z_arr = arithmetic_op(y_arr, x_arr)
+    z = arithmetic_op(y, x)
+
+    assert all_almost_equal([x, y, z], [x_arr, y_arr, z_arr], places=places)
+
+    # aliased operation
+    z_arr = arithmetic_op(x_arr, x_arr)
+    z = arithmetic_op(x, x)
+
+    assert all_almost_equal([x, y, z], [x_arr, y_arr, z_arr], places=places)
+
+
+def test_assign(tspace):
+    """Test the assign method using ``==`` comparison."""
+    x = noise_element(tspace)
+    y = noise_element(tspace)
+
+    y.assign(x)
+
+    assert y == x
+    assert y is not x
+
+    # test alignment
+    x *= 2
+    assert y != x
+
+
+def test_inner(tspace):
+    """Test the inner method against numpy.vdot."""
+    xd = noise_element(tspace)
+    yd = noise_element(tspace)
+
+    correct_inner = np.vdot(yd, xd)
```

Comment:
This only works for unweighted spaces, which implies that the `tspace` fixture only returns unweighted spaces. Fix and add to the fixture.